3-dynamic-analysis/1-debugging/1-delve.md

```diff
@@ -0,0 +1,171 @@
+# Debugging Go programs with delve
+
+For quite a long time people asked for a good debugger for Go.
+GDB works, but it doesn't really understand Go routines and that causes complications
+when debugging Go programs as soon as they start to get a bit complicated.
+
+Luckily, the Go community provided a new debugger called [delve](https://github.com/derekparker/delve).
+If you haven't installed it yet, please follow [these instructions](https://github.com/derekparker/delve/tree/master/Documentation/installation)
+before continuing.
+
+We will debug the web server implemented in [this directory](../webserver).
+It's a very simple web server that checks the path of the HTTP request and it
+prints `hello, foo` if the path was something like `localhost:8080/foo@golang.org`,
+and `hello, stranger` otherwise.
+
+Try it now. It might not work ... and that's ok. Do not fix it just yet!
+
+## Debugging the webserver
+
+To start the `webserver` program simply visit the directory with your terminal
+and run:
+
+```
+$ dlv debug
+Type 'help' for list of commands.
+(dlv)
+```
+
+If you try to visit the http://localhost:8080 you will see that the server is not running.
+This is normal, since our binary was loaded but not executed yet! Let's make it run by
+typing `continue`.
+
+```
+$ dlv debug
```

Comment:
It is magical! I will add it if people ask for it. I think that it's good you don't need to understand all of this extra knowledge other than `dlv debug` generates  a binary that can be debugged